### DISCUSSION TITLE
Is there any way to change 1 key to auto-type date?

### DISCUSSION THREAD

**Discussion by user283287427** [pqrs-org/Karabiner-Elements#3710] [Q&A]
Hi, I'd like to auto-type today's date when I click the button under the Esc button, because I don't use this symbol.

Edit: I found this code and it works but pretty slow (in a delay). Do you think there's a better way to achieve that?

```
{
    "description": "hyper + d inserts datetime",
    "manipulators": [
        {
            "from": {
                "key_code": "non_us_backslash"
            },
            "to": [
                {
                    "shell_command": "osascript -e \"tell application \\\"System Events\\\" to keystroke \\\"$(date $@ +'%Y-%m-%d')\\\"\""
                }
            ],
            "type": "basic"
        }
    ]
}

```

**Comment by MaxFreedomPollard** [comment]
The delay is mostly the `osascript` route: every press spawns osascript, which compiles the AppleScript and then drives System Events to type the string one character at a time. Two ways to make it snappy.

**Option A: type it with `cliclick` instead of osascript (my pick).**

`cliclick` is a tiny binary that starts much faster than osascript and types text directly, so it skips the AppleScript overhead that's costing you the delay.

```bash
brew install cliclick   # then grant it Accessibility permission
```

```json
{
    "description": "insert today's date",
    "manipulators": [
        {
            "from": { "key_code": "non_us_backslash" },
            "to": [
                { "shell_command": "/opt/homebrew/bin/cliclick t:\"$(date +%Y-%m-%d)\"" }
            ],
            "type": "basic"
        }
    ]
}
```

Use the absolute path from `which cliclick`, because Karabiner runs `shell_command` with a bare PATH. This doesn't touch the clipboard and is about as close to instant as you'll get.

**Option B: copy + paste (one keystroke instead of ten).**

If you'd rather not add a dependency, put the date on the clipboard and paste it. Pasting is a single Cmd+V rather than typing each character:

```json
{
    "from": { "key_code": "non_us_backslash" },
    "to": [
        { "shell_command": "printf '%s' \"$(date +%Y-%m-%d)\" | pbcopy && osascript -e 'tell application \"System Events\" to keystroke \"v\" using command down'" }
    ],
    "type": "basic"
}
```

Keep the copy and the paste inside the **same** `shell_command` joined with `&&`. Karabiner runs shell commands asynchronously, so if you split the paste into a separate `to` key event, it can fire before the clipboard is set. The `&&` guarantees the order. Downside is it overwrites whatever was on your clipboard.

I'd start with A. The AppleScript/System Events path is where the lag comes from, and cliclick sidesteps it entirely.

(Verified the `date` output and the `pbcopy`/`pbpaste` round-trip on macOS; cliclick's `t:` types into the frontmost app.)
